timelock-corruption-detection/src/main/java/com/palantir/history/remote/HistoryLoaderAndTransformer.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.history.remote;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+
+import com.google.common.collect.Maps;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.history.LocalHistoryLoader;
+import com.palantir.history.models.LearnerAndAcceptorRecords;
+import com.palantir.history.models.PaxosHistoryOnSingleNode;
+import com.palantir.paxos.NamespaceAndUseCase;
+import com.palantir.timelock.history.HistoryQuery;
+import com.palantir.timelock.history.LogsForNamespaceAndUseCase;
+import com.palantir.timelock.history.PaxosLogWithAcceptedAndLearnedValues;
+
+public final class HistoryLoaderAndTransformer {
+    private HistoryLoaderAndTransformer() {
+        //no op
+    }
+
+    public static List<LogsForNamespaceAndUseCase> getLogsForHistoryQueries(LocalHistoryLoader localHistoryLoader,
+            List<HistoryQuery> historyQueries) {
+        Map<NamespaceAndUseCase, Long> lastVerifiedSequences = historyQueries.stream().collect(
+                Collectors.toMap(HistoryQuery::getNamespaceAndUseCase, HistoryQuery::getSeq, Math::min));
+
+        PaxosHistoryOnSingleNode localPaxosHistory = localHistoryLoader.getLocalPaxosHistory(lastVerifiedSequences);
+
+        return KeyedStream.stream(localPaxosHistory.history())
+                .mapEntries(HistoryLoaderAndTransformer::processHistory)
+                .values()
+                .collect(Collectors.toList());
+    }
+
+    private static Map.Entry<NamespaceAndUseCase, LogsForNamespaceAndUseCase> processHistory(
+            NamespaceAndUseCase namespaceAndUseCase, LearnerAndAcceptorRecords records) {
+
+        long minSeq = records.getMinSequence();
+        long maxSeq = records.getMaxSequence();
+
+        List<PaxosLogWithAcceptedAndLearnedValues> logs = LongStream.rangeClosed(minSeq, maxSeq).boxed().map(
+                sequence -> PaxosLogWithAcceptedAndLearnedValues.builder()
+                        .paxosValue(records.getLearnedValueAtSeqIfExists(sequence))
+                        .acceptedState(records.getAcceptedValueAtSeqIfExists(sequence))
+                        .seq(sequence)
+                        .build()).collect(Collectors.toList());
```

Comment:
Can we iterate over the map entries, as opposed to over this range? I'm a bit worried about this at deployments like big internal stack, where we truncated a good number of these entries so we have -1 and then a big gap, and then four million.